File: java-e2e/src/test/java/com/testflow/e2e/pages/LoginPage.java
package com.testflow.e2e.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;

import static com.microsoft.playwright.assertions.PlaywrightAssertions.assertThat;

/**
 * POM for TestFlow's auth modal (there is no dedicated /login route — sign-in is a modal
 * triggered by the "Sign in" button on any page). Mirrors {@code e2e/pages/login.page.ts}.
 */
public class LoginPage extends BasePage {

    private final Locator heading;
    private final Locator signInHeading;
    private final Locator usernameInput;
    private final Locator passwordInput;
    private final Locator submitBtn;
    private final Locator authError;
    private final Locator contactAdminText;

    public LoginPage(Page page) {
        super(page);
        this.heading = page.locator("#auth-form-container h1", new Page.LocatorOptions().setHasText("TestFlow"));
        this.signInHeading = page.locator("#auth-form-container h2", new Page.LocatorOptions().setHasText("Sign in"));
        this.usernameInput = page.getByTestId("auth-username");
        this.passwordInput = page.getByTestId("auth-password");
        this.submitBtn = page.getByTestId("auth-submit-btn");
        this.authError = page.locator("#auth-error");
        this.contactAdminText = page.getByText("Contact your admin to get an account.");
    }

    /** Loads the app as a guest and opens the sign-in modal. */
    public LoginPage open() {
        navigate("/");
        waitForNetworkIdle();
        page.getByTestId("signin-btn").click();
        submitBtn.waitFor();
        return this;
    }

    /** Opens the sign-in modal and submits credentials (does not assume success). */
    public LoginPage login(String username, String password) {
        open();
        usernameInput.fill(username);
        passwordInput.fill(password);
        submitBtn.click();
        return this;
    }

    public LoginPage expectModalLoaded() {
        heading.waitFor();
        signInHeading.waitFor();
        usernameInput.waitFor();
        passwordInput.waitFor();
        submitBtn.waitFor();
        return this;
    }

    public boolean contactAdminTextVisible() {
        return contactAdminText.isVisible();
    }

    public String loginErrorText() {
        authError.waitFor();
        return authError.textContent();
    }

    public LoginPage expectLoggedIn() {
        assertThat(page.getByTestId("logout-btn")).isVisible();
        return this;
    }
}

File: e2e-bdd/pages/login.page.ts
import { Page, Locator, expect } from '@playwright/test';
import { BasePage } from './base.page';

/**
 * POM for TestFlow's auth modal — the BDD suite's own copy, kept independent
 * from ../e2e/pages/login.page.ts on purpose (see e2e-bdd/README.md): every
 * automation stack in this repo (pytest, Playwright TS, REST Assured, Java
 * Playwright) already owns its Page Objects rather than sharing them, and
 * that independence also sidesteps a real TypeScript conflict — two
 * separately-installed @playwright/test copies produce structurally
 * incompatible `Page` types, so importing across the package boundary
 * doesn't type-check cleanly.
 */
export class LoginPage extends BasePage {
  readonly heading:          Locator;
  readonly signInHeading:    Locator;
  readonly usernameInput:    Locator;
  readonly passwordInput:    Locator;
  readonly submitBtn:        Locator;
  readonly authError:        Locator;
  readonly contactAdminText: Locator;

  constructor(page: Page) {
    super(page);
    this.heading          = page.locator('#auth-form-container h1', { hasText: 'TestFlow' });
    this.signInHeading    = page.locator('#auth-form-container h2', { hasText: 'Sign in' });
    this.usernameInput    = page.getByTestId('auth-username');
    this.passwordInput    = page.getByTestId('auth-password');
    this.submitBtn        = page.getByTestId('auth-submit-btn');
    this.authError        = page.locator('#auth-error');
    this.contactAdminText = page.getByText('Contact your admin to get an account.');
  }

  async open(): Promise<void> {
    await this.navigate('/');
    await this.waitForNetworkIdle();
    await this.page.getByTestId('signin-btn').click();
    await expect(this.submitBtn).toBeVisible();
  }

  async login(username: string, password: string): Promise<void> {
    await this.open();
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.submitBtn.click();
  }

  async expectModalLoaded(): Promise<void> {
    await expect(this.heading).toBeVisible();
    await expect(this.signInHeading).toBeVisible();
    await expect(this.usernameInput).toBeVisible();
    await expect(this.passwordInput).toBeVisible();
    await expect(this.submitBtn).toBeVisible();
  }

  async expectLoginError(text: string): Promise<void> {
    await expect(this.authError).toBeVisible();
    await expect(this.authError).toContainText(text);
  }

  async expectLoggedIn(): Promise<void> {
    await expect(this.page.getByTestId('logout-btn')).toBeVisible();
  }
}
